File: etapa3-construcao(geracao-de-cod)/stack1/Hospede.ts
import { HospedeData, ValidationResult } from '../../types';
import { CPF } from '../value-objects/CPF';
import { Email } from '../value-objects/Email';
import { v4 as uuidv4 } from 'uuid';

export class Hospede {
  readonly id: string;
  private _nome: string;
  private _sobrenome: string;
  private _cpf: CPF;
  private _email: Email;
  readonly criadoEm: Date;
  private _atualizadoEm: Date;

  constructor(params: {
    nome: string;
    sobrenome: string;
    cpf: CPF;
    email: Email;
    id?: string;
    criadoEm?: Date;
    atualizadoEm?: Date;
  }) {
    this.id = params.id || uuidv4();
    this._nome = params.nome;
    this._sobrenome = params.sobrenome;
    this._cpf = params.cpf;
    this._email = params.email;
    this.criadoEm = params.criadoEm || new Date();
    this._atualizadoEm = params.atualizadoEm || new Date();
  }

  get nome(): string {
    return this._nome;
  }

  get sobrenome(): string {
    return this._sobrenome;
  }

  get cpf(): CPF {
    return this._cpf;
  }

  get email(): Email {
    return this._email;
  }

  get atualizadoEm(): Date {
    return this._atualizadoEm;
  }

  nomeCompleto(): string {
    return `${this._nome} ${this._sobrenome}`;
  }

  atualizarEmail(email: Email): void {
    this._email = email;
    this._atualizadoEm = new Date();
  }

  atualizarDados(params: {
    nome?: string;
    sobrenome?: string;
    email?: Email;
  }): void {
    if (params.nome !== undefined) this._nome = params.nome;
    if (params.sobrenome !== undefined) this._sobrenome = params.sobrenome;
    if (params.email !== undefined) this._email = params.email;
    
    this._atualizadoEm = new Date();
  }

  validar(): ValidationResult {
    const errors: string[] = [];

    if (!this._nome || this._nome.trim().length === 0) {
      errors.push('Nome é obrigatório');
    }

    if (!this._sobrenome || this._sobrenome.trim().length === 0) {
      errors.push('Sobrenome é obrigatório');
    }

    return {
      isValid: errors.length === 0,
      errors
    };
  }

  toData(): HospedeData {
    return {
      id: this.id,
      nome: this._nome,
      sobrenome: this._sobrenome,
      cpf: this._cpf.valorSemFormatacao(),
      email: this._email.getValor(),
      criadoEm: this.criadoEm,
      atualizadoEm: this._atualizadoEm
    };
  }

  static fromData(data: HospedeData): Hospede {
    return new Hospede({
      nome: data.nome,
      sobrenome: data.sobrenome,
      cpf: CPF.criar(data.cpf),
      email: Email.criar(data.email),
      id: data.id,
      criadoEm: data.criadoEm,
      atualizadoEm: data.atualizadoEm
    });
  }
}

File: etapa3-construcao(geracao-de-cod)/stack2/src/backend/src/main/java/com/hotel/reserva/domain/entities/Hospede.java
package com.hotel.reserva.domain.entities;

import com.hotel.reserva.domain.valueobjects.CPF;
import com.hotel.reserva.domain.valueobjects.Email;
import jakarta.persistence.*;
import lombok.*;

import java.time.LocalDateTime;
import java.util.UUID;

@Entity
@Table(name = "hospedes")
@Getter
@NoArgsConstructor(access = AccessLevel.PROTECTED)
@EqualsAndHashCode(of = "id")
public class Hospede {

    @Id
    @GeneratedValue(strategy = GenerationType.UUID)
    private UUID id;

    @Column(nullable = false)
    private String nome;

    @Column(nullable = false)
    private String sobrenome;

    @Embedded
    @AttributeOverride(name = "valor", column = @Column(name = "cpf", nullable = false, unique = true))
    private CPF cpf;

    @Embedded
    @AttributeOverride(name = "valor", column = @Column(name = "email", nullable = false))
    private Email email;

    @Column(nullable = false, updatable = false)
    private LocalDateTime criadoEm;

    @Column(nullable = false)
    private LocalDateTime atualizadoEm;

    @PrePersist
    protected void onCreate() {
        criadoEm = LocalDateTime.now();
        atualizadoEm = LocalDateTime.now();
    }

    @PreUpdate
    protected void onUpdate() {
        atualizadoEm = LocalDateTime.now();
    }

    @Builder
    public Hospede(String nome, String sobrenome, String cpf, String email) {
        this.nome = nome;
        this.sobrenome = sobrenome;
        this.cpf = CPF.criar(cpf);
        this.email = Email.criar(email);
    }

    public String getNomeCompleto() {
        return nome + " " + sobrenome;
    }

    public void atualizarEmail(Email novoEmail) {
        this.email = novoEmail;
        this.atualizadoEm = LocalDateTime.now();
    }

    public void atualizarDados(String nome, String sobrenome, String email) {
        if (nome != null && !nome.isBlank()) {
            this.nome = nome;
        }
        
        if (sobrenome != null && !sobrenome.isBlank()) {
            this.sobrenome = sobrenome;
        }
        
        if (email != null && !email.isBlank()) {
            this.email = Email.criar(email);
        }
        
        this.atualizadoEm = LocalDateTime.now();
    }

    public void validar() {
        if (nome == null || nome.isBlank()) {
            throw new IllegalArgumentException("Nome é obrigatório");
        }
        
        if (sobrenome == null || sobrenome.isBlank()) {
            throw new IllegalArgumentException("Sobrenome é obrigatório");
        }
    }
}
